python/pyarrow/_csv.pyx

```diff
@@ -0,0 +1,200 @@
+        if block_size is not None:
+            self.block_size = block_size
+        if num_rows is not None:
+            self.num_rows = num_rows
+
+    @property
+    def use_threads(self):
+        return self.options.use_threads
+
+    @use_threads.setter
+    def use_threads(self, value):
+        self.options.use_threads = value
+
+    @property
+    def block_size(self):
+        return self.options.block_size
+
+    @block_size.setter
+    def block_size(self, value):
+        self.options.block_size = value
+
+    @property
+    def num_rows(self):
+        return self.options.num_rows
+
+    @num_rows.setter
+    def num_rows(self, value):
+        self.options.num_rows = value
+
+
+cdef class CSVParseOptions:
+    cdef:
+        CCSVParseOptions options
+
+    __slots__ = ()
+
+    def __init__(self, delimiter=None, quote_char=None, double_quote=None,
+                 escape_char=None, header_rows=None):
+        self.options = CCSVParseOptions.Defaults()
+        if delimiter is not None:
+            self.delimiter = delimiter
+        if quote_char is not None:
+            self.quote_char = quote_char
+        if double_quote is not None:
+            self.double_quote = double_quote
+        if escape_char is not None:
+            self.escape_char = escape_char
+        if header_rows is not None:
+            self.header_rows = header_rows
+
+    @property
+    def delimiter(self):
+        return chr(self.options.delimiter)
+
+    @delimiter.setter
+    def delimiter(self, value):
+        self.options.delimiter = _single_char(value)
+
+    @property
+    def quote_char(self):
+        if self.options.quoting:
+            return chr(self.options.quote_char)
+        else:
+            return False
+
+    @quote_char.setter
+    def quote_char(self, value):
+        if value is False:
+            self.options.quoting = False
+        else:
+            self.options.quote_char = _single_char(value)
+            self.options.quoting = True
+
+    @property
+    def double_quote(self):
+        return self.options.double_quote
+
+    @double_quote.setter
+    def double_quote(self, value):
+        self.options.double_quote = value
+
+    @property
+    def escape_char(self):
+        if self.options.escaping:
+            return chr(self.options.escape_char)
+        else:
+            return False
+
+    @escape_char.setter
+    def escape_char(self, value):
+        if value is False:
+            self.options.escaping = False
+        else:
+            self.options.escape_char = _single_char(value)
+            self.options.escaping = True
+
+    @property
+    def header_rows(self):
+        return self.options.header_rows
+
+    @header_rows.setter
+    def header_rows(self, value):
+        self.options.header_rows = value
+
+
+cdef _get_reader(input_file, shared_ptr[InputStream]* out):
+    cdef shared_ptr[RandomAccessFile] result
+    use_memory_map = False
+    get_reader(input_file, use_memory_map, &result)
```

Comment:
The problem is that ReadaheadSpooler will make a copy anyway when reading from a memory-mapped file (because of padding).

More generally, I don't think it is really relevant here. Parsing CSV data will always be much slower than the speed of raw memory copies (which should be on the order of 30 GB/s from main memory).

Reply:
By the way, another motivation for not focussing on memory-mapped files is that it wouldn't help with compressed CSV files.